underfs/cos/src/main/java/alluxio/underfs/cos/COSUnderFileSystem.java

```diff
@@ -0,0 +1,310 @@
+  @Override
+  public void setMode(String path, short mode) throws IOException {}
+
+  @Override
+  protected boolean copyObject(String src, String dst) {
+    try {
+      LOG.info("Copying {} to {}", src, dst);
+      mClient.copyObject(mBucketNameInternal, src, mBucketNameInternal, dst);
+      return true;
+    } catch (CosClientException e) {
+      LOG.error("Failed to rename file {} to {}", src, dst, e);
+      return false;
+    }
+  }
+
+  @Override
+  protected boolean createEmptyObject(String key) {
+    try {
+      ObjectMetadata objMeta = new ObjectMetadata();
+      objMeta.setContentLength(0);
+      mClient.putObject(mBucketNameInternal, key, new ByteArrayInputStream(new byte[0]), objMeta);
+      return true;
+    } catch (CosClientException e) {
+      LOG.error("Failed to create object: {}", key, e);
+      return false;
+    }
+  }
+
+  @Override
+  protected OutputStream createObject(String key) throws IOException {
+    return new COSOutputStream(mBucketNameInternal, key, mClient);
+  }
+
+  @Override
+  protected boolean deleteObject(String key) {
+    try {
+      mClient.deleteObject(mBucketNameInternal, key);
+    } catch (CosClientException e) {
+      LOG.error("Failed to delete {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected String getFolderSuffix() {
+    return FOLDER_SUFFIX;
+  }
+
+  @Override
+  protected ObjectListingChunk getObjectListingChunk(String key, boolean recursive)
+      throws IOException {
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
+    key = PathUtils.normalizePath(key, PATH_SEPARATOR);
+    // In case key is root (empty string) do not normalize prefix
+    key = key.equals(PATH_SEPARATOR) ? "" : key;
+    ListObjectsRequest request = new ListObjectsRequest();
+    request.setBucketName(mBucketNameInternal);
+    request.setPrefix(key);
+    request.setMaxKeys(getListingChunkLength());
+    request.setDelimiter(delimiter);
+
+    ObjectListing result = getObjectListingChunk(request);
+    if (result != null) {
+      return new COSObjectListingChunk(request, result);
+    }
+    return null;
+  }
+
+  // Get next chunk of listing result
+  private ObjectListing getObjectListingChunk(ListObjectsRequest request) {
+    ObjectListing result;
+    try {
+      result = mClient.listObjects(request);
+    } catch (CosClientException e) {
+      LOG.error("Failed to list path {}", request.getPrefix(), e);
+      result = null;
+    }
+    return result;
+  }
+
+  /**
+   * Wrapper over COS {@link ObjectListingChunk}.
+   */
+  private final class COSObjectListingChunk implements ObjectListingChunk {
+    final ListObjectsRequest mRequest;
+    final ObjectListing mResult;
+
+    COSObjectListingChunk(ListObjectsRequest request, ObjectListing result) throws IOException {
+      mRequest = request;
+      mResult = result;
+      if (mResult == null) {
+        throw new IOException("COS listing result is null");
```

Comment:
suggest: to enforce `result` is not null, instead of throwing `IOException`, Alluxio codebase typically adds the following precondition check and avoid throwing IOE on constructors:

```java
Preconditions.checkNotNull(result, "result");
```